File: Data Structures & Algorithms/balanced-binary-tree/submission-0.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def isBalanced(self, root: Optional[TreeNode]) -> bool:
        

        def dfs(root):
            if not root:
                return [True, 0]

            # left and right

            left = dfs(root.left)
            right = dfs(root.right)

            isBalanced = True

            if (abs(left[1] - right[1]) > 1 or left[0] != True or right[0] != True):
                isBalanced = False
            
            return [isBalanced, max(left[1], right[1]) + 1]

        return dfs(root)[0]

File: Data Structures & Algorithms/invert-a-binary-tree/submission-0.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    public TreeNode invertTree(TreeNode root) {
        // Create another TreeNode and having the opposite left and right node
        // Traverse the tree, append the origial node with opposite left and right to the new tree

        if (root == null){
            return null;
        }
        
        TreeNode newRoot = new TreeNode(root.val);
        
        newRoot.right = invertTree(root.left);
        newRoot.left = invertTree(root.right);

        return newRoot;
    }
}
